cvat/apps/engine/task.py

```diff
@@ -319,68 +321,101 @@ def update_progress(progress):
+            if manifest_file and not os.path.exists(db_data.get_manifest_path()):
+                shutil.copyfile(os.path.join(upload_dir, manifest_file[0]),
+                    db_data.get_manifest_path())
+                if upload_dir != settings.SHARE_ROOT:
+                    os.remove(os.path.join(upload_dir, manifest_file[0]))
+
             if task_mode == MEDIA_TYPES['video']['mode']:
                 try:
-                    if meta_info_file:
+                    if manifest_file:
                         try:
-                            from cvat.apps.engine.prepare import UploadedMeta
-                            meta_info = UploadedMeta(source_path=os.path.join(upload_dir, media_files[0]),
-                                                     meta_path=db_data.get_meta_path(),
-                                                     uploaded_meta=os.path.join(upload_dir, meta_info_file[0]))
-                            meta_info.check_seek_key_frames()
-                            meta_info.check_frames_numbers()
-                            meta_info.save_meta_info()
-                            assert len(meta_info.key_frames) > 0, 'No key frames.'
+                            manifest = VideoManifestValidator(source_path=os.path.join(upload_dir, media_files[0]),
+                                                              manifest_path=db_data.get_manifest_path())
+                            manifest.init_index()
+                            manifest.validate_seek_key_frames()
+                            manifest.validate_frame_numbers()
+                            assert len(manifest) > 0, 'No key frames.'
+
+                            all_frames = manifest['properties']['length']
+                            video_size = manifest['properties']['resolution']
                         except Exception as ex:
                             base_msg = str(ex) if isinstance(ex, AssertionError) else \
                                 'Invalid meta information was upload.'
-                            job.meta['status'] = '{} Start prepare valid meta information.'.format(base_msg)
-                            job.save_meta()
+                            _update_status('{} Start prepare valid meta information.'.format(base_msg))
                             meta_info, smooth_decoding = prepare_meta(
+                                data_type='video',
                                 media_file=media_files[0],
                                 upload_dir=upload_dir,
-                                meta_dir=os.path.dirname(db_data.get_meta_path()),
                                 chunk_size=db_data.chunk_size
                             )
                             assert smooth_decoding == True, 'Too few keyframes for smooth video decoding.'
+                            _update_status('Start prepare a manifest file')
+                            manifest = VideoManifestManager(db_data.get_manifest_path())
```

Comment:
Nice redesign. But now, as I can see, there are 2 duplicating code blocks in `except` and `else`, and I think, it's worth trying to leave only one of them.